vm/compiler/src/ledger/store/block/mod.rs

```diff
@@ -83,6 +83,116 @@ pub trait BlockStorage<N: Network>: Clone + Sync {
     /// Returns the signature map.
     fn signature_map(&self) -> &Self::SignatureMap;
 
+    /// Starts an atomic batch write operation.
+    fn start_atomic(&self) {
+        self.id_map().start_atomic();
+        self.reverse_id_map().start_atomic();
+        self.header_map().start_atomic();
+        self.transactions_map().start_atomic();
+        self.reverse_transactions_map().start_atomic();
+        self.transaction_store().start_atomic();
+        self.signature_map().start_atomic();
+    }
+
+    /// Aborts an atomic batch write operation.
+    fn abort_atomic(&self) {
+        self.id_map().abort_atomic();
+        self.reverse_id_map().abort_atomic();
+        self.header_map().abort_atomic();
+        self.transactions_map().abort_atomic();
+        self.reverse_transactions_map().abort_atomic();
+        self.transaction_store().abort_atomic();
+        self.signature_map().abort_atomic();
+    }
+
+    /// Finishes an atomic batch write operation.
+    fn finish_atomic(&self) {
+        self.id_map().finish_atomic();
+        self.reverse_id_map().finish_atomic();
+        self.header_map().finish_atomic();
+        self.transactions_map().finish_atomic();
+        self.reverse_transactions_map().finish_atomic();
+        self.transaction_store().finish_atomic();
+        self.signature_map().finish_atomic();
+    }
+
+    /// Stores the given `block` into storage.
+    fn insert(&self, block: &Block<N>) -> Result<()> {
+        // Start an atomic batch write operation.
+        self.start_atomic();
+
+        // Store the block hash.
+        self.id_map().insert(block.height(), block.hash()).or_abort(|| self.abort_atomic())?;
+        // Store the block height.
+        self.reverse_id_map().insert(block.hash(), block.height()).or_abort(|| self.abort_atomic())?;
+        // Store the block header.
+        self.header_map().insert(block.hash(), *block.header()).or_abort(|| self.abort_atomic())?;
+
+        // Store the transaction IDs.
+        self.transactions_map()
+            .insert(block.hash(), block.transaction_ids().copied().collect())
+            .or_abort(|| self.abort_atomic())?;
+
+        // Store the block transactions.
+        for transaction in block.transactions().values() {
+            // Store the reverse transaction ID.
+            self.reverse_transactions_map().insert(transaction.id(), block.hash()).or_abort(|| self.abort_atomic())?;
+            // Store the transaction.
+            self.transaction_store().insert(transaction).or_abort(|| self.abort_atomic())?;
+        }
+
+        // Store the block signature.
+        self.signature_map().insert(block.hash(), *block.signature()).or_abort(|| self.abort_atomic())?;
+
+        // Finish the atomic batch write operation.
+        self.finish_atomic();
+
+        Ok(())
+    }
```

Comment:
alternatively, these functions could look as follows (example for `BlockStorage::insert`):
```
/// Stores the given `block` into storage.
fn insert(&self, block: &Block<N>) -> Result<()> {
    // Start an atomic batch write operation.
    self.start_atomic();

    let atomic_ops = || -> Result<()> {
        // Store the block hash.
        self.id_map().insert(block.height(), block.hash())?;
        // Store the block height.
        self.reverse_id_map().insert(block.hash(), block.height())?;
        // Store the block header.
        self.header_map().insert(block.hash(), *block.header())?;

        // Store the transaction IDs.
        self.transactions_map().insert(block.hash(), block.transaction_ids().copied().collect())?;

        // Store the block transactions.
        for transaction in block.transactions().values() {
            // Store the reverse transaction ID.
            self.reverse_transactions_map().insert(transaction.id(), block.hash())?;
            // Store the transaction.
            self.transaction_store().insert(transaction)?;
        }

        // Store the block signature.
        self.signature_map().insert(block.hash(), *block.signature())?;

        Ok(())
    };

    if atomic_ops().is_err() {
        self.abort_atomic();
    }

    // Finish the atomic batch write operation.
    self.finish_atomic();

    Ok(())
}
```

Reply:
Agreed! Would you mind updating each of these cases to this approach? I think it may be easier to checkout an earlier commit and hash and add the wrapping closure.

Reply:
Lets tackle this as a follow up